### ISSUE TITLE
K2 defaults should conform to community large cluster config

### ISSUE THREAD

**Issue by countspongebob** [samsung-cnct/kraken-lib#620]
Community documentation here:

https://github.com/kubernetes/community/blob/master/sig-scalability/provider-configs.md

For the control plane config:
1) Split etcd
2) 5 node config 
3) Note instances sizes specified

This issue is specific to the control plane config, not the number of nodes brought up, which could also use some discussion.

**Comment by countspongebob** [comment]
@lostlivio @jandrake  Opened per our conversation.

**Comment by davidewatson** [comment]
1. and 2. are currently the case, however, the default k2 configuration uses `m4.large` for master nodes instead of `m4.16xlarge`. That equates to an 8x increase in memory and a 16x increase in cores.

I am not sure the defaults should assume such a large k8s installation. Our CI system is overwhelmingly the largest consumer of k2 defaults, and in that case, there are only 10 worker nodes. Instead, it might be faster and cheaper to have a weekly scalability CI test which specifically tests the proposed configuration.

**Comment by countspongebob** [comment]
The default I'm requesting is smaller than the current default.

Alternatively, a single-node non-HA system would be reasonable.

**Comment by countspongebob** [comment]
@davidewatson Now I see your concern. My primary issue was the size of the control plane config, not the default number of started nodes.

**Comment by davidewatson** [comment]
Another area k2 defaults differ from the SIG scale testing proposal is 100GB vs. 256 GB etcd storage partitions for etcd.

**Comment by countspongebob** [comment]
The 5 node cluster is total. Etcd configs share the same 5 nodes but the storage is partitioned for iops isolation. Total of 5 nodes for all control plane. @davidewatson do you want to open issues for these specific issues?

**Comment by countspongebob** [comment]
@joejulian @sostheim Can you please comment based on customer requirements you are aware of? Note that customers should not be named in this thread.

**Comment by venezia** [comment]
Since we know etcd has a fixed ceiling in size, I feel 100gb might be way too big?

Mike

**Comment by davidewatson** [comment]
@countspongebob: Currently, k2 provisioned etcd servers must exist within a single k2 nodePool. Allowing multiple etcd instances to share a single node might be substantially easier once we support self-hosted k8s (e.g. #153 and #168), although, that is quite a bit of work, and may not even be acceptable for users large enough to require split etcd.

@venezia: The 256GB number was determined during SIG scale based on the number of IOPS required.

@coffeepac: We'll need to decide how to prioritize this.

**Comment by countspongebob** [comment]
Iops are pegged to size only on GCE, on AWS you could use a smaller partition with committed iops.

**Comment by countspongebob** [comment]
For clarity, the Kubernetes community doc is the target here. The doc above was studied and referenced in the development of the target configuration.

**Comment by sostheim** [comment]
This topic (etcd node size) actually came up today in working with a staging cluster and the planned production cluster.  A few thoughts:
* There is sensitivity to the cost of the cloud provider infrastructure used for dev/staging clusters.
* Production deployments are unlikely to use these defaults except in the rarest of cases where a maxima cluster is desired.
* Specifically, the production deployment targets are very much in line with the coreos hardware guide linked in the previous comment.
* An active area for cluster tuning is in the individual sizing of the: 
1. load balanced API servers
1. etcd cluster state store
1. etcd cluster event store
Each of these has a very different performance characteristic and should be sized independently.  The number of API servers is a factor of client network traffic and is severable from the number of etcd nodes for the cluster state store.
The number of etcd nodes in the cluster state and event processing clusters is clearly unrelated.

**Comment by sostheim** [comment]
If I could summarize the all of the above simply, the defaults need to evaluated for every cluster deployment.  

It's fair to assume that whatever the defaults selected for k2 are, they need to be modified for every cluster deployment.

**Comment by countspongebob** [comment]
I believe the current K2 defaults are oversized. The scalability defaults (community) are smaller and more reasonable. For many clients just sizing those nodes should get them in the right zone.

**Comment by countspongebob** [comment]
Agree, but K2 needs to support colocating the split etcd instances and iiops partitioning strategies.

ALTERNATIVELY the default could be a really small non-ha cluster, but I like the default representing community best practice in HA/scalable clusters.

**Comment by leahnp** [comment]
@coffeepac am I right in understanding this will be a "breaking config file change"?

**Comment by leahnp** [comment]
My understanding of what we want out of this ticket, taking into considerations the k8s large cluster doc, coreos etcd specs, this conversation and the current kraken config:

- Increase master node size to: m4.16xlarge 
- Etcd version bump to: 3.1.8 (currently 3.1.0)
- Run etcd and etcd-events in the same node (5, m4.16xlarge nodes) - Etcd and Etcd-events will mount to separate storage & run in separate RKT containers
- Up etcd and etcd-events storage volume size from 100 to 256

One question - should we wait for @davidewatson 's etcd container work to be done before proceeding with the unification of etcd and etcd-event nodes?

**Comment by davidewatson** [comment]
In regards to "support [for] colocating the split etcd instances and [IOPS] partitioning strategies."

tl;dr The schema should be fine, but we will want to test k2 upgrades.

We can avoid making this a breaking schema change by allowing the existing `etcdConfig` property to accept either an `etcdConfig`, or a list of `etcdConfig`s. If we do this, a comment should be added to the v2alpha ticket noting that in the next version there should be only a single property `etcdConfigs`, with a single type of list.

In order to allow different `etcdConfig`s to target different partitions, we will also need a way to point a particular etcd instance at a particular directory. This is to provide IOPS isolation. Note that the mount points are defined in `nodeConfig`.

The larger concern I have is whether this is a breaking change for existing users. Some testing will be needed (in particular I have not considered `k2cli update`), but I think we may be okay since we disallow reconfiguration of etcd after `up`.

**Comment by leahnp** [comment]
Have the config mostly in the right state to adhere to these standards. Working through tweaking terraform today. Hope to get to the final component (modding: k2/ansible/roles/kraken.nodePool/kraken.nodePool.etcd/templates/etcd.units.etcd.part.jinja2) soon. 

Ectd v3.2.5 upgrade passed CI.

**Comment by leahnp** [comment]
Backwards compatibility code added, parameterized drunkensmee scripts, need to test. Dryrun is running successfully.

**Comment by leahnp** [comment]
Meeting with @davidewatson & @alejandroEsc today to scope this ticket, established to not cover following issues:

- backwards compatibility for etcdConfig
- refactoring extra loops in: kraken.provider.aws.module.etcd.tf.jinja2

I am going to close #781 as I feel I am too far in the weeds, I will be opening a new PR with this newly scoped work.

**Comment by leahnp** [comment]
- going to save moving certs to: /etc/etcd/etcd-name/... for another PR, would require too many changes to make this PR feasible.

**Comment by leahnp** [comment]
Etcd nodes will spin up but, currently failing at:
`core@ip-10-0-69-54 /etc $ systemctl list-units --state=failed
  UNIT                          LOAD   ACTIVE SUB    DESCRIPTION              
● kraken-etcd-ssl.service       loaded failed failed Setup Network Environment
● kraken-etcdEvents-ssl.service loaded failed failed Setup Network Environment`

possibly related to location of certs (etc/etcd/etcd-name/...) updating drunkensmee pathing to see if this is where the error is originating.

**Comment by leahnp** [comment]
Still on same bug, some approaches I have tried to find the problem:

- comparing user-data with working etcd node
- comparing unit files with working etcd node
- examining unit files/status of services that kraken-etcdEvents-ssl.service `require` and `after` on
- examining RKT logs, only `setup-network-environment` available in state `exited garbage` (the garbage is concerning as it is different than a working nodes state of `exited)
- drunkensmee never started indicating it's failing on `Setup Network Environment`

**Comment by leahnp** [comment]
@davidewatson found issue: must pass variables to shell script with `--` prepended (script in drunkensmee)

**Comment by leahnp** [comment]
ETCD nodes seem to be up and healthy. Apiserver currently crashing. Working on solution, currently investigating "RBAC DENY" error and "TLS TIMEOUT" (both of which also occur in healthy master nodes/apiserver containers) but the TLS TIMEOUT is followed by a golang stacktrace that might indicate the problem.

**Comment by leahnp** [comment]
Reverting work to use same certs for etcd and etcdEvents - Kubernetes api only supports single etcd certs (due to `--etcd-cafile=` flag).

**Comment by leahnp** [comment]
Added backwards compatibility for configs with "etcdConfig" https://github.com/samsung-cnct/kraken-lib/pull/858

**Comment by davidewatson** [comment]
Instead of making 16xlarge the default for all clusters, it would be better if we provided an option to generate different default configs for different purposes (e.g. AWS vs. GKE, small, medium, and large clusters, etc.).

Some reasons include:

 - Additional documentation and manual actions will be required for all users.
 - The default AWS instance limit for 16xlarge's is 5. Since the default etcd cluster size is 5, if we make this the default then users of kraken will not be able to create more than one cluster at a time. It may also complicate recovery since a new instance cannot be provisioned until the failed instance is destroyed.
 - 45x (= 15 * 5 etcd nodes) cost increase for each developer, user, and customer.
 - Increased boot time.

For large scale cluster testing we need a separate CI job, probably running density e2e tests. This is similar to what Kubernetes does.

@countspongebob: Please review this comment to ensure development meets the spirit of the issue.

**Comment by jandrake** [comment]
After reviewing all of the above, my view is this:

- Best practices for HA control plane for kubernetes is the goal
- HA and Load are separate beasts, but related
- Kraken should default to HA control plane CONFIGURATION always (5 nodes, deployed as described)
- Kraken should allow for choice in sizing and provide scaling options based on proven testing
- It should be very feasible for a dev to run a low node count Kraken cluster (and this may help with our CI challenges as well)
- CNCT should own expertise in control plane SCALING which is translated to each provider appropriately.  This is a matrix test suite of provider and scalability thresholds
- CNCT should have test suites demonstrating these SCALING configurations and be run regularly as part of Kraken release processes.
- Kraken should never be deployed with less than a five node control plane as our focus is solely on production clusters

Likely there are some detail regarding etcd scaling and risks as described in the community standard for cascading problems the may inform the above.  Happy to hear them.

**Comment by davidewatson** [comment]
@jandrake: Kraken clusters are configured for HA, and according to Best Practices, by default.

There are three different changes recommended by this issue.

1) Etcd instances should have sufficient memory and CPU resources. For 5k node clusters, the etcd instances should be m4.16xlarge.
2) etcd instances should have sufficient IOPs. For 5k node clusters on GCP, this can be ensured by giving each etcd instance a 256 GB partition.
3) Split (logical) etcd clusters should share the same hardware (https://github.com/samsung-cnct/kraken-lib/pull/858).

2) and 3) are in progress. I claim we do satisfy 1) for most clusters with our current default instance size of m4.xlarge, and therefore shouldn't make m4.16xlarge the default since most clusters do not have 5k nodes. If we want to make it the default anyway, we can. It's just problematic for the reasons I enumerated above.

I think our current defaults are a good starting point for building a ClusterOps ready k8s cluster. I recommend anyone creating a production deployment consider whether the defaults are optimal for their use case.

**Comment by jandrake** [comment]
@dwat I think we're in violent agreement here.  I'm actually agreeing with your prior comments about instance sizes.  Obviously, Kraken users can manually modify instance sizes for etcd clusters.  I am going one step further and suggesting that we take your recommendation and formalize it.  Specifically, if we know m4.xlarge service 5K clusters, we should know various size clusters all the way down to single developer size clusters in tested variations that make sense to us.  I would advocate a kraken flag for number of expected nodes and/or use cases.  I would advocate for us to have as part of our release tests full regression for each size of cluster we create a flag for.  I would expect that the two flags supported initially would be "5K production" and "development" conceptually (don't care about what the flags specifically are).

HA != Capacity.  HA = architecture and instancing as related to load.  Anyway, dead horse.  Stopping beating it.

**Comment by leahnp** [comment]
Working on issues raised in: https://github.com/samsung-cnct/drunkensmee/pull/2
Currently configuring drunkensmee to write individual peer certs for each etcd instance on node. Seeing error:
```cluster may be unhealthy: failed to list members
Error:  client: etcd cluster is unavailable or misconfigured
error #0: dial tcp 10.0.86.196:2379: getsockopt: connection refused
error #1: dial tcp 10.0.180.226:2379: getsockopt: connection refused
error #2: dial tcp 10.0.63.202:2379: getsockopt: connection refused``` but I am confused how clientport (2379) is being affected as I am under the impression I have only altered peer certs and peer cert locations.

**Comment by leahnp** [comment]
Updates to: https://github.com/samsung-cnct/drunkensmee/pull/2 in working order.

**Comment by leahnp** [comment]
Drunkensmee PR review resulted in choice to swap drunkensmee repo to new team standard: https://github.com/samsung-cnct/drunkensmee/issues/4

**Comment by leahnp** [comment]
Created new drunkensmee: https://github.com/samsung-cnct/container-drunkensmee
Another commit to the Kraken-lib PR - this should be close once we get the new image of drunkensmee to replace my dev image.

**Comment by leahnp** [comment]
https://github.com/samsung-cnct/container-drunkensmee working and Jenkins pushing image to quay.

Started giant rebase of kraken-lib work to pull in new dict commit

Got a successful CI run for kraken-lib
